api/v1_storage.inc

```diff
@@ -0,0 +1,32 @@
+<?php
+
+// DP API v1 -- Storage
+
+//===========================================================================
+// storage/clients/:clientid
+
+/** @return mixed */
+function api_v1_storage_clients(string $method, array $data, array $query_params)
+{
+    global $pguser;
+
+    $clientid = $data[":clientid"];
+
+    $storage = new ApiClientStorage($clientid, $pguser);
+    if ($method == "GET") {
+        return json_decode($storage->get());
+    } elseif ($method == "PUT") {
+        $storage->set(json_encode(api_get_request_body()));
```

Comment:
In this case it is already encoded by the client so we don't want to encode it again.

Reply:
> In fact, given the existing API structure, we do want to encode it twice when returning the string to the client. The client encodes its javascript object into a JSON string and expects to get back the same string.

No, that's not it works.

The Client isn't sending us an opaque string, it's sending us a JSON object. That comes through as a string and the ApiRouter decodes that string into an object -- it does this for all request bodies and it how all of the functions work. Our ApiClientStorage `get()` and `set()`s take strings, so we have to encode that back into a string to store it.

Client sends JSON in request -> API converts to object -> `v1_storage.inc` converts back to string -> ApiClientStorage persists string

When we `get()` it from ApiClientStorage the class returns a string. But returns from API router functions needs to be an object, because the API will encode that into a string, so we need to decode it into an object first.

ApiClientStorage returns string -> `v1_storage.inc` converts back to object -> API converts object to string -> Client receives JSON in return

Reply:
> p.s. I think the api_get_request_body() is a bit confusing because when it says return $json, the thing it is returning generally isn't JSON but what was originally provided as data to the ajax function. (though in my scenario it now is).

Yes, that's totally fair -- `api_get_request_body()` returns a decoded object/array, not a string - `$json_object` would be clearer.

And yes, the client is encoding the object into JSON -- we do that already in the ajax function:
```js
    if (upperCaseMethod !== "GET") {
        // POST or PUT
        options.headers["Content-Type"] = "application/json";
        options.body = JSON.stringify(data);
    }
```
The client just needs to pass in the object it wants persisted as `$data` to `ajax()` and it'll all just work.

Reply:
I see what you mean. But I think there could be some problems with that approach. The original javascript object will get encoded into JSON and will get decoded into a PHP object and then re-encoded into JSON and the reverse on the way back. PHP objects don't correspond exactly to javascript objects. As we have seen, a php associative array which happens to have consecutive numerical keys or is empty is encoded as an ordinary array. What we get back may not be the same as what we started with.
The other approach of having the client make a JSON object with JSONstringify and then sending that (it would get doubly encoded) would, I imagine, be safer since we are only json-encoding and -decoding strings in the server.  v1_storage would then not need to do any encoding or decoding but the client would have to encode and decode so the number of codings would end up much the same overall.

Reply:
If the client wants to send an encoded JSON string as part of a JSON object they're more than welcome to do so (`options.body = {data = JSON.stringify(data)};` or however the JS syntax is) although I think it's a bad idea, interoperability issues or not.

It's unsafe and unwise for us to store completely unstructured blobs which is what you're asking for.

Reply:
If what the client sends is not valid JSON then the database will reject it.

Reply:
So somehow bypass the API infra? I'm willing to entertain it if you want to fork my branch and show me what you have in mind.

Reply:
No, I don't want to do that. When we save a page from the PI we send a string from the client and the same string gets put into the db. What I'm proposing is sending a JSON string in the same way and this same JSON string gets put into the db as  JSON.

Reply:
The page is a string inside a JSON object -- it's not a naked string. So what you're proposing is having a JSON string as a value of a JSON object:
```json
{
    "data": "{\"key\": \"value\"}"
}
```
And then the API would extract `data` out of the payload and pass the string into the ApiClientStorage class.

That's a weird and unexpected interface. No one does that -- just pass the JSON object 😁

Reply:
It's getting late here now. I'll consider it later but have a happy Christmas.

Reply:
I've put up a commit that allows a routed function to get the non-deserialized input and output non-serialized output. This is less hacky than I thought it would be. It's in a separate commit for evaluation / discussion.

Reply:
Essentially yes but without the wrapping.